core/primitives/benches/benches.rs

```diff
@@ -12,48 +12,62 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-// TODO: Move benchmark to criterion #2354
-#![feature(test)]
 
-extern crate test;
-use hex_literal::hex;
+#[macro_use]
+extern crate criterion;
+
+use criterion::{Criterion, black_box, Bencher, Fun};
+use std::time::Duration;
 use substrate_primitives::hashing::{twox_128, blake2_128};
 
 const MAX_KEY_SIZE: u32 = 32;
 
-fn data_set() -> Vec<Vec<u8>> {
+fn get_key(key_size: u32) -> Vec<u8> {
 	use rand::SeedableRng;
 	use rand::Rng;
 
 	let rnd: [u8; 32] = rand::rngs::StdRng::seed_from_u64(12).gen();
 	let mut rnd = rnd.iter().cycle();
-	let mut res = Vec::new();
-	for size in 1..=MAX_KEY_SIZE {
-		for _ in 0..1_000 {
-			let value = (0..size)
-				.map(|_| rnd.next().unwrap().clone())
-				.collect();
-			res.push(value);
-		}
-	}
-	res
+
+	(0..key_size)
+		.map(|_| rnd.next().unwrap().clone())
+		.collect()
+}
+
+fn bench_blake2_128(b: &mut Bencher, key: &Vec<u8>) {
+	b.iter(|| {
+		let _a = blake2_128(black_box(key));
+	});
 }
 
-fn bench_hash_128(b: &mut test::Bencher, f: &Fn(&[u8]) -> [u8; 16]) {
-	let data_set = data_set();
+fn bench_twox_128(b: &mut Bencher, key: &Vec<u8>) {
 	b.iter(|| {
-		for data in &data_set {
-			let _a = f(data);
-		}
+		let _a = twox_128(black_box(key));
 	});
 }
 
-#[bench]
-fn bench_blake2_128(b: &mut test::Bencher) {
-	bench_hash_128(b, &blake2_128);
+fn bench_hash_128_fix_size(c: &mut Criterion) {
+	let key = get_key(MAX_KEY_SIZE);
+	let blake_fn = Fun::new("blake2_128", bench_blake2_128);
+	let twox_fn = Fun::new("twox_128", bench_twox_128);
+	let fns = vec![blake_fn, twox_fn];
+
+	c.bench_functions("fixed size hashing", fns, key);
+}
+
+fn bench_hash_128_dyn_size(c: &mut Criterion) {
+	let mut keys = Vec::new();
+	for i in (2..MAX_KEY_SIZE).step_by(4) {
```

Comment:
`(2..MAX_KEY_SIZE).step_by(4).map(get_key).collect()`

Just some minor tip for the future.. :D